### DISCUSSION TITLE
CXX Bridge Question (problem with dependency cycle)

### DISCUSSION THREAD

**Discussion by felixguendling** [corrosion-rs/corrosion#489] [Q&A] [answered]
Hey! :wave: C++/Rust interop newbie here.

I am experimenting to make functionality from a C++ library available in Rust (e.g. to write a HTTP API for the C++ code in Rust). This works fine so far. However, I guess I made a mistake since I needed to manually copy the `cxx.h` file into the C++ project to break a dependency cycle.

- So there is an existing C++ library (here: `nigiri`) that I rather not want to modify.
- Then I have some wrapper code (also in C++): here `nigiri-rust`. This already requires the `cxx.h` include since it uses `rust::*` types.
- CXX / corosion then generates another library (here `nigiri-cxx`), which is then linked into the Rust executable (`srv`) via: `corrosion_link_libraries(srv nigiri-cxx nigiri-rust)`

The full change set is here:
https://github.com/motis-project/nigiri/compare/master...rust

Is this basically the right way?

I would have liked to pull in the `cxx.h` file in `nigiri-rust` from the generated `nigiri-cxx`. However, this is not possible because `nigiri-cxx` also requires `nigiri-rust` for includes. So CMake rightfully complains about a cycle.

How can I get writ of this cycle and have a codebase that's free of `cxx.h`. As I understand it, this should only be generated (probably to stay up-to-date with the CXX version defined in `Cargo.toml`).

**Comment by felixguendling** [comment, marked as answer]
Figured it out based on this test:
https://github.com/corrosion-rs/corrosion/blob/master/test/cxxbridge/cxxbridge_cpp2rust/CMakeLists.txt

I think it would be helpful to add a link in the book.

**Reply by jschwe** [reply]
a pull-request improving the documentation or linking the test would be very welcome.
